app/src/main/java/org/onionshare/android/ui/MainActivity.kt

```diff
@@ -1,55 +1,130 @@
+import org.onionshare.android.server.WebserverManager.State.STOPPED
+import org.onionshare.android.server.WebserverManager.State.STOPPING
 import org.onionshare.android.ui.theme.OnionshareTheme
+import org.onionshare.android.ui.theme.PurpleOnionShare
+import org.slf4j.LoggerFactory.getLogger
 
 @AndroidEntryPoint
 class MainActivity : ComponentActivity() {
+
+    private val viewModel: MainViewModel by viewModels()
+
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContent {
             OnionshareTheme {
                 // A surface container using the 'background' color from the theme
                 Surface(color = MaterialTheme.colors.background) {
-                    Greeting("OnionSharer")
+                    Greeting(
+                        name = "OnionSharer",
+                        webserverState = viewModel.webserverState,
+                        onButtonClicked = this::onButtonClicked,
+                        onFabClicked = this::onFabClicked,
+                    )
                 }
             }
         }
     }
+
+    private fun onButtonClicked(currentState: State) = when (currentState) {
+        STOPPED -> viewModel.startServer()
+        STARTED -> viewModel.stopServer()
+        else -> error("Illegal click state: $currentState")
+    }
+
+    private val launcher = registerForActivityResult(GetMultipleContents()) { uris ->
+        // not supporting selecting entire folders with sub-folders
+        uris.forEach { uri ->
+            getLogger("TEST").error("$uri")
+        }
+        Toast.makeText(this, "Not implemented", LENGTH_SHORT).show()
+    }
+
+    private fun onFabClicked() {
+        launcher.launch("*/*")
+    }
 }
 
 @Composable
-fun Greeting(name: String) {
-    Box(modifier = Modifier
-        .fillMaxWidth()
-        .fillMaxHeight()
+fun Greeting(
+    name: String,
+    webserverState: StateFlow<State>,
+    onButtonClicked: (State) -> Unit,
+    onFabClicked: () -> Unit,
+) {
+    Scaffold(
+        floatingActionButton = {
+            val text = "Add files"
+            ExtendedFloatingActionButton(
+                onClick = onFabClicked,
+                icon = { Icon(Icons.Filled.Add, contentDescription = text) },
+                text = { Text(text) },
+                backgroundColor = PurpleOnionShare,
+            )
+        },
     ) {
-        Text(text = "Hello $name!", modifier = Modifier
-            .padding(all = Dp(16.0f))
-            .align(Alignment.Center)
-        )
+        Column(
+            horizontalAlignment = Alignment.CenterHorizontally,
+            verticalArrangement = Arrangement.Center,
+            modifier = Modifier
+                .fillMaxWidth()
+                .fillMaxHeight()
+        ) {
+            Text(
+                text = "Hello $name!",
+                modifier = Modifier.padding(all = Dp(16.0f))
+            )
+            val state = webserverState.collectAsState().value
```

Comment:
I will now go and do some reading about Compose...

Reply:
Yeah that'd be good. I am also new to this and don't know if this is the right way to do it.